### PR TITLE
Auto-populated link descriptions and thumbnails

### PR DESCRIPTION
**Pull request by wassgha**
Partial to #253 

Demo of auto link metadata population

### Changes
- Implements automated populated of link metadata (title, description and thumbnail) while adding a two-tap link
- Fixes an issue that prevented links from being saved

### CODE REVIEW

**Review by spacedmonkey — DISMISSED** [review]
Travis is failing. Will review when these tests pass.

File: package.json [hunk 1]
```diff
@@ -40,6 +40,7 @@
     "resize-observer-polyfill": "^1.5.1",
     "styled-components": "^5.0.1",
     "throttle-debounce": "^2.1.0",
```
Comment (spacedmonkey) [inline package.json]: Remove `throttle-debounce` ?

File: assets/src/edit-story/components/panels/link.js [hunk 2]
```diff
@@ -82,29 +84,40 @@ function LinkPanel({ selectedElements, onSetProperties }) {
...
-  const populateMetadata = useCallback(
-    debounce(300, async (/** url */) => {
-      // TODO(wassgha): Implement getting the page metadata
-    })
-  );
+
+  const [populateMetadata] = useDebouncedCallback((url) => {
+    setFetchingMetadata(true);
+    getLinkMetadata(url).then(({ title, image }) => {
+      setState((originalState) => ({
```
Comment (miina) [inline assets/src/edit-story/components/panels/link.js]: Should this do anything differently when the link is actually incorrect and data was not received?

File: includes/REST_API/Link_Controller.php [hunk 3]
```diff
@@ -131,6 +131,11 @@ public function parse_link( $request ) {
 
 		$response = wp_safe_remote_get( $url, $args );
 
+		$redirect_location = $response['headers']['location'];
```
Comment (miina) [inline includes/REST_API/Link_Controller.php]: This will cause a fatal error if `$response` is an error.

Comment (wassgha) [reply includes/REST_API/Link_Controller.php]: I'll leave this to another PR

File: assets/src/edit-story/components/panels/link.js [hunk 4]
```diff
@@ -82,29 +85,43 @@ function LinkPanel({ selectedElements, onSetProperties }) {
...
+
   const canLink = selectedElements.length === 1 && !isFill;
-  const populateMetadata = useCallback(
-    debounce(300, async (/** url */) => {
-      // TODO(wassgha): Implement getting the page metadata
-    })
-  );
+
+  const [populateMetadata] = useDebouncedCallback((url) => {
+    if (!validUrl.isUri(url)) {
```
Comment (dvoytenko) [inline assets/src/edit-story/components/panels/link.js]: What would this say about "www.example.com/"?

File: assets/src/edit-story/components/panels/link.js [hunk 5]
```diff
@@ -82,29 +85,43 @@ function LinkPanel({ selectedElements, onSetProperties }) {
...
+      return;
+    }
+    setFetchingMetadata(true);
+    getLinkMetadata(url).then(({ title, image }) => {
```
Comment (dvoytenko) [inline assets/src/edit-story/components/panels/link.js]: Let's also immediately assume that URL itself can change. E.g. "http://" can be switched to "https://".

**Review by dvoytenko — CHANGES_REQUESTED** [review]

File: assets/src/edit-story/components/panels/link.js (hunk 4 above)
Comment (wassgha) [reply assets/src/edit-story/components/panels/link.js]: The scheme is required, do we want to allow `www.example.com/` ? If so, can we assume https?

File: assets/src/edit-story/components/panels/link.js (hunk 5 above)
Comment (wassgha) [reply assets/src/edit-story/components/panels/link.js]: We don't get that information from the current API, I'll make an issue for @swissspidy  to look into

Comment (wassgha) [reply assets/src/edit-story/components/panels/link.js]: @jauyong could you please triage this and make an issue accordingly? The task is to return the redirected url from the PHP API for retrieving url metadata.

File: assets/src/edit-story/components/panels/link.js (hunk 4 above)
Comment (pbakaus) [reply assets/src/edit-story/components/panels/link.js]: yeah I'd say let's allow unprefixed, and assume https:// then.

File: assets/src/edit-story/components/panels/link.js (hunk 5 above)
Comment (spacedmonkey) [reply assets/src/edit-story/components/panels/link.js]: There are some nice utils for [urls](https://github.com/WordPress/gutenberg/tree/master/packages/url/src) in gutenberg - 
Might be useful here. Specially [prepend-http](https://github.com/WordPress/gutenberg/blob/master/packages/url/src/prepend-http.js)

File: assets/src/edit-story/components/panels/link.js (hunk 4 above)
Comment (spacedmonkey) [reply assets/src/edit-story/components/panels/link.js]: There are some nice utils for [urls](https://github.com/WordPress/gutenberg/tree/master/packages/url/src) in gutenberg - 
Might be useful here. Specially [prepend-http](https://github.com/WordPress/gutenberg/blob/master/packages/url/src/prepend-http.js)

Comment (wassgha) [reply assets/src/edit-story/components/panels/link.js]: Already implemented my own but I'll keep an eye on that if we need something more complex! Thanks for the reference!

File: assets/src/edit-story/components/panels/link.js [hunk 6]
```diff
@@ -82,29 +85,48 @@ function LinkPanel({ selectedElements, onSetProperties }) {
...
+      return;
+    }
+    setFetchingMetadata(true);
+    getLinkMetadata(withProtocol(url))
+      .then(({ title, image }) => {
+        setState((originalState) => ({
+          ...originalState,
+          desc: title ? title : originalState.desc,
+          icon: image
+            ? toAbsoluteUrl(withProtocol(originalState.url), image)
```
Comment (dvoytenko) [inline assets/src/edit-story/components/panels/link.js]: Not sure what this is meant to do, but visibly it looks odd: first it adds the protocol and then tries to make an absolute URL. But a URL with a protocol is always absolute. No?

**Review by dvoytenko — CHANGES_REQUESTED** [review]

File: assets/src/edit-story/components/panels/link.js (hunk 6 above)
Comment (wassgha) [reply assets/src/edit-story/components/panels/link.js]: It tries to make `image` an absolute URL, not `originalState.url`. `image` can sometimes be `/images/something.jpg` which btw I think should be solved on the API side, but till then this should work

Comment (dvoytenko) [reply assets/src/edit-story/components/panels/link.js]: I agree: it should definitely be an API. But at this point I got your point. We can move with this and come back to it on API level.

**Review by dvoytenko — APPROVED** [review]

File: assets/src/edit-story/utils/url.js [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+export { isUri as isValidUrl } from 'valid-url';
+export { resolve as toAbsoluteUrl } from 'url';
```
Comment (dvoytenko) [inline assets/src/edit-story/utils/url.js]: Let's just use `new URL(...)`

**Review by dvoytenko — APPROVED** [review]
LG with one request

**Pull request merged**